src/components/DappBrowser/handleProviderRequest.ts

```diff
@@ -0,0 +1,342 @@
+  tab?: Tab | undefined;
+  /** The tabs.Tab which opened the connection, if any. This property will only be present when the connection was opened from a tab (including content scripts), and only if the receiver is an extension, not an app. */
+  title?: string;
+}
+
+export type CallbackOptions = {
+  /** The sender of the message. */
+  sender: IMessageSender;
+  /** The topic provided. */
+  topic: string;
+  /** An optional scoped identifier. */
+  id?: number | string;
+};
+
+/**
+ * Creates a generic transport that can be used to send and receive messages between extension scripts
+ * under a given topic & set of types.
+ *
+ * @see https://www.notion.so/rainbowdotme/Cross-script-Messaging-141de5115294435f95e31b87abcf4314#3b63e155df6a4b71b0e6e74f7a2c416b
+ */
+export function createTransport<TPayload, TResponse>({ messenger, topic }: { messenger: Messenger; topic: string }) {
+  if (!messenger.available) {
+    console.error(`Messenger "${messenger.name}" is not available in this context.`);
+  }
+  return {
+    async send(payload: TPayload, { id }: { id: number }) {
+      return messenger.send<TPayload, TResponse>(topic, payload, { id });
+    },
+    async reply(callback: (payload: TPayload, callbackOptions: CallbackOptions) => Promise<TResponse>) {
+      messenger.reply(topic, callback);
+    },
+  };
+}
+
+/**
+ * Uses extensionMessenger to send messages to popup for the user to approve or reject
+ * @param {PendingRequest} request
+ * @returns {boolean}
+ */
+const messengerProviderRequestFn = async (messenger: Messenger, request: ProviderRequestPayload) => {
+  const appSession = appSessionsStore.getState().getActiveSession({ host: getDappHost(request.meta?.sender.url) || '' });
+
+  // Wait for response from the popup.
+  let response: unknown | null;
+
+  if (request.method === 'eth_requestAccounts') {
+    response = await handleDappBrowserConnectionPrompt({
+      dappName: request.meta?.sender.title || '',
+      dappUrl: request.meta?.sender.url || '',
+    });
+
+    appSessionsStore.getState().addSession({
+      host: getDappHost(request.meta?.sender.url) || '',
+      // @ts-ignore
+      address: response.address,
+      // @ts-ignore
+      network: getNetworkFromChainId(response.chainId),
+      // @ts-ignore
+      url: request.meta?.sender.url || '',
+    });
+  } else {
+    response = await handleDappBrowserRequest({
+      dappName: request.meta?.sender.title || request.meta?.sender.url || '',
+      imageUrl: '',
+      address: appSession?.address || '',
+      network: appSession?.network || Network.mainnet,
+      dappUrl: request.meta?.sender.url || '',
+      payload: request,
+    });
+  }
+
+  if (!response) {
+    throw new UserRejectedRequestError(Error('User rejected the request.'));
+  }
+  return response;
+};
+
+const isSupportedChainId = (chainId: number) => RainbowNetworks.filter(network => Number(network.id) === chainId).length > 0;
```

Comment:
i feel like this should be

const isSupportedChainId = (chainId: number) => !!RainbowNetworks.find(network => Number(network.id) === chainId);

Reply:
I feel like there shouldn't be networks at all but ok

Reply:
insert spiderman meme